File: examples/Delphi/Table/Table 11 GetCellAttributes/ExTbl11U.pas
{$J+} {Writable constants}

unit ExTbl11U;

interface

uses
  SysUtils, Windows, Messages, Classes, Graphics, Controls,
  Forms, Dialogs, OvcEf, OvcPb, OvcPf, OvcTCPic, OvcTCBEF, OvcTCHdr,
  OvcTCmmn, OvcTable, OvcTCStr, OvcTCEdt, OvcBase, StdCtrls, Buttons,
  OvcTCell;

type
  TForm1 = class(TForm)
    OvcTable1: TOvcTable;
    OvcController1: TOvcController;
    OvcTCString1: TOvcTCString;
    OvcTCString2: TOvcTCString;
    OvcTCRowHead1: TOvcTCRowHead;
    OvcTCPictureField1: TOvcTCPictureField;
    BitBtn1: TBitBtn;
    procedure OvcTable1GetCellData(Sender: TObject; RowNum: Longint;
      ColNum: Integer; var Data: Pointer; Purpose : TOvcCellDataPurpose);
    procedure OvcTable1GetCellAttributes(Sender: TObject; RowNum: Longint;
      ColNum: Integer; var CellAttr: TOvcCellAttributes);
    procedure BitBtn1Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}


procedure TForm1.OvcTable1GetCellData(Sender: TObject; RowNum: Longint;
  ColNum: Integer; var Data: Pointer; Purpose : TOvcCellDataPurpose);
type
  TTestRec = record
    Name  : string;
    Age   : word;
  end;
  TTestDataBase = array [1..9] of TTestRec;
const
  TestDatabase : TTestDataBase = (
    (Name:'DelRossi, Robert'; Age:36),
    (Name:'Blanton, Phillip'; Age:33),
    (Name:'Reisdorph, Kent';  Age:25),
    (Name:'Welch, Mike';      Age:31),
    (Name:'Frerking, Gary';   Age:27),
    (Name:'Jones, Zack';      Age:35),
    (Name:'Bucknall, Julian'; Age:38),
    (Name:'';                 Age:0),
    (Name:'';                 Age:0));
const
  ColHeading : string = '';
begin
  Data := nil;
  if (1 <= RowNum) and (RowNum <= 9) then
    begin
      case ColNum of
        1 : Data := @TestDatabase[RowNum].Name;
        2 : Data := @TestDatabase[RowNum].Age;
      end;{case}
    end
  else
    begin
      case ColNum of
        0 : ColHeading := 'TP';
        1 : ColHeading := 'Name';
        2 : ColHeading := 'Age';
      end;{case}
      Data := @ColHeading;
    end;
end;

procedure TForm1.OvcTable1GetCellAttributes(Sender: TObject;
  RowNum: Longint; ColNum: Integer; var CellAttr: TOvcCellAttributes);
begin
  if (RowNum = 0) and (ColNum = 2) then
    CellAttr.caAdjust := otaCenterRight;
end;

procedure TForm1.BitBtn1Click(Sender: TObject);
begin
  Close;
end;

end.

File: examples/CBuildr5/ExTbl11U.cpp
//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include "ExTbl11U.h"
//---------------------------------------------------------------------------
#pragma link "OvcTable"
#pragma link "OvcTCmmn"
#pragma link "OvcBase"
#pragma link "OvcTCEdt"
#pragma link "OvcTCStr"
#pragma link "OvcTCell"
#pragma link "OvcTCHdr"
#pragma link "OvcTCPic"
#pragma link "OvcTCBEF"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
  TestDatabase[0].Name = "Bucknall, Julian";
  TestDatabase[0].Age = 39;
  TestDatabase[1].Name = "Hughes, Terry";
  TestDatabase[1].Age = 37;
  TestDatabase[2].Name = "Inman, Lee";
  TestDatabase[2].Age = 44;
  TestDatabase[3].Name = "Reisdorph, Kent";
  TestDatabase[3].Age = 36;
  TestDatabase[4].Name = "Leier, John";
  TestDatabase[4].Age = 33;
  TestDatabase[5].Name = "Frerking, Gary";
  TestDatabase[5].Age = 33;
  TestDatabase[6].Name = "Rose, Larry";
  TestDatabase[6].Age = 36;
  TestDatabase[7].Name = "";
  TestDatabase[7].Age = 0;
  TestDatabase[8].Name = "";
  TestDatabase[8].Age = 0;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcTable1GetCellData(TObject *Sender, int RowNum,
  int ColNum, Pointer &Data, TOvcCellDataPurpose Purpose)
{
  Data = NULL;
  if (1 <= RowNum && RowNum <= 9) {
    switch (ColNum) {
      case 1 : Data = &TestDatabase[RowNum - 1].Name; break;
      case 2 : Data = &TestDatabase[RowNum - 1].Age; break;
    }
  }
  else {
    switch (ColNum) {
      case 0 : ColHeading = "TP"; break;
      case 1 : ColHeading = "Name"; break;
      case 2 : ColHeading = "Age"; break;
    }
    Data = &ColHeading;
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcTable1GetCellAttributes(TObject *Sender, int RowNum,
  int ColNum, TOvcCellAttributes &CellAttr)
{
  if ((RowNum == 0) && (ColNum == 2))
    CellAttr.caAdjust = otaCenterRight;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::BitBtn1Click(TObject *Sender)
{
  Close();    
}
//---------------------------------------------------------------------------

